#VRML_SIM R2023b utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/floors/protos/RectangleArena.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/factory/containers/protos/CardboardBox.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/factory/containers/protos/WoodenBox.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/factory/pipes/protos/PipeSection.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/factory/tools/protos/Wrench.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/factory/tools/protos/Hammer.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/factory/valves/protos/LeverValve.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/factory/conveyors/protos/ConveyorBelt.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/buildings/protos/Warehouse.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/buildings/protos/SimpleBuilding.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/road/protos/RoadIntersection.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/traffic/protos/TrafficCone.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/obstacles/protos/OilBarrel.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/geometries/protos/Cylinder.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/geometries/protos/Box.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/street_furniture/protos/Fence.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/CorrugatedMetal.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/BrushedAluminium.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/Concrete.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/RustyMetal.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/humans/pedestrian/protos/Pedestrian.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/robots/unitree/go2/protos/Go2.proto"

WorldInfo {
  info [
    "Construction Site Simulation with Unitree Go2 Robot"
    "Realistic environment with human workers, construction equipment, and safety barriers"
    "Author: AI Assistant"
  ]
  title "Construction Site Environment"
  basicTimeStep 16
  contactProperties [
    ContactProperties {
      material1 "robot_material"
      material2 "ground_material"
      coulombFriction [
        0.8
      ]
      bounce 0.1
      bounceVelocity 0.01
    }
    ContactProperties {
      material1 "human_material"
      material2 "ground_material"
      coulombFriction [
        0.9
      ]
      bounce 0.0
      bounceVelocity 0.01
    }
  ]
}

Viewpoint {
  orientation -0.3 0.85 0.42 1.2
  position 25 15 20
  follow "go2_robot"
}

TexturedBackground {
  texture "factory"
}

TexturedBackgroundLight {
}

# Main construction site floor with proper boundaries
RectangleArena {
  translation 0 0 0
  floorSize 50 50
  floorTileSize 2 2
  wallHeight 0.5
  wallThickness 0.1
  floorAppearance Concrete {
    colorOverride 0.7 0.7 0.6
  }
  wallAppearance CorrugatedMetal {
  }
}

# Unitree Go2 Robot with proper sensors
Go2 {
  translation 0 0 0.3
  rotation 0 0 1 0
  name "go2_robot"
  controller "webots_robot_controller"
  supervisor TRUE
  customData "construction_site_patrol"
  bodySlot [
    # Additional sensors can be added here
    Camera {
      translation 0.2 0 0.1
      rotation 0 1 0 0
      name "front_camera"
      width 640
      height 480
      fieldOfView 1.047
    }
    RangeFinder {
      translation 0.2 0 0.1
      rotation 0 1 0 0
      name "depth_camera"
      width 640
      height 480
      fieldOfView 1.047
      maxRange 10
    }
    Lidar {
      translation 0 0 0.15
      rotation 0 0 1 0
      name "lidar"
      horizontalResolution 512
      fieldOfView 6.28
      verticalFieldOfView 0.5
      numberOfLayers 16
      maxRange 20
      noise 0.01
    }
  ]
}

# Human Workers with proper movement patterns
Pedestrian {
  translation 10 8 1.27
  rotation 0 0 1 1.57
  name "worker_1"
  controller "human_controller"
  controllerArgs [
    "worker_1"
  ]
  shirtColor 0.2 0.6 1.0
  pantsColor 0.3 0.3 0.3
  shoesColor 0.1 0.1 0.1
  skinColor 0.9 0.7 0.6
}

Pedestrian {
  translation -8 12 1.27
  rotation 0 0 1 -1.57
  name "worker_2"
  controller "human_controller"
  controllerArgs [
    "worker_2"
  ]
  shirtColor 1.0 0.5 0.0
  pantsColor 0.2 0.2 0.8
  shoesColor 0.2 0.1 0.0
  skinColor 0.8 0.6 0.4
}

Pedestrian {
  translation 5 -10 1.27
  rotation 0 0 1 0
  name "worker_3"
  controller "human_controller"
  controllerArgs [
    "worker_3"
  ]
  shirtColor 0.0 0.8 0.2
  pantsColor 0.1 0.1 0.1
  shoesColor 0.3 0.2 0.1
  skinColor 0.95 0.8 0.7
}

# Construction Building Under Construction
SimpleBuilding {
  translation 15 10 0
  floorNumber 2
  floorHeight 3.5
  startingFloor 0
  corners [
    8 8
    8 -8
    -8 -8
    -8 8
  ]
  wallType "glass highrise"
  roofShape "flat roof"
}

# Warehouse for materials storage
Warehouse {
  translation -15 -10 0
  rotation 0 0 1 1.57
}

# Construction Equipment and Materials

# Material Storage Area
WoodenBox {
  translation 8 15 0.3
  name "material_box_1"
  size 2 1 0.6
}

WoodenBox {
  translation 10 15 0.3
  name "material_box_2"
  size 2 1 0.6
}

CardboardBox {
  translation 12 15 0.15
  name "supply_box_1"
}

CardboardBox {
  translation 14 15 0.15
  name "supply_box_2"
}

# Pipe sections for construction
PipeSection {
  translation 6 18 0.1
  rotation 0 0 1 1.57
  name "pipe_1"
  height 0.2
  radius 0.1
  thickness 0.01
  subdivision 16
}

PipeSection {
  translation 8 18 0.1
  rotation 0 0 1 1.57
  name "pipe_2"
  height 0.2
  radius 0.1
  thickness 0.01
  subdivision 16
}

# Construction Tools
Wrench {
  translation 5 16 0.6
  name "wrench_1"
}

Hammer {
  translation 7 16 0.6
  name "hammer_1"
}

# Industrial Equipment
LeverValve {
  translation -10 5 0.5
  name "valve_1"
}

ConveyorBelt {
  translation -5 -15 0
  rotation 0 0 1 0
  name "conveyor_1"
  length 8
  width 1
  speed 0.2
}

# Safety Equipment and Barriers

# Safety Cones
TrafficCone {
  translation 2 2 0
  name "cone_1"
}

TrafficCone {
  translation -2 2 0
  name "cone_2"
}

TrafficCone {
  translation 2 -2 0
  name "cone_3"
}

TrafficCone {
  translation -2 -2 0
  name "cone_4"
}

# Oil barrels for industrial storage
OilBarrel {
  translation 18 -5 0.44
  name "barrel_1"
}

OilBarrel {
  translation 20 -5 0.44
  name "barrel_2"
}

OilBarrel {
  translation 22 -5 0.44
  name "barrel_3"
}

# Safety Barriers
Fence {
  translation 0 20 0
  rotation 0 0 1 0
  name "barrier_north"
  height 1.2
  length 15
  numberOfPoles 6
}

Fence {
  translation 20 0 0
  rotation 0 0 1 1.57
  name "barrier_east"
  height 1.2
  length 15
  numberOfPoles 6
}

Fence {
  translation 0 -20 0
  rotation 0 0 1 3.14
  name "barrier_south"
  height 1.2
  length 15
  numberOfPoles 6
}

Fence {
  translation -20 0 0
  rotation 0 0 1 -1.57
  name "barrier_west"
  height 1.2
  length 15
  numberOfPoles 6
}

# Scaffolding Structure (Custom built)
DEF SCAFFOLDING Group {
  children [
    # Vertical poles
    Cylinder {
      translation 12 5 1.5
      name "scaffold_pole_1"
      height 3
      radius 0.05
      appearance BrushedAluminium {
      }
    }
    Cylinder {
      translation 12 8 1.5
      name "scaffold_pole_2"
      height 3
      radius 0.05
      appearance BrushedAluminium {
      }
    }
    Cylinder {
      translation 15 5 1.5
      name "scaffold_pole_3"
      height 3
      radius 0.05
      appearance BrushedAluminium {
      }
    }
    Cylinder {
      translation 15 8 1.5
      name "scaffold_pole_4"
      height 3
      radius 0.05
      appearance BrushedAluminium {
      }
    }
    # Horizontal platforms
    Box {
      translation 13.5 6.5 2.5
      name "scaffold_platform_1"
      size 3 3 0.1
      appearance RustyMetal {
      }
    }
    Box {
      translation 13.5 6.5 1.5
      name "scaffold_platform_2"
      size 3 3 0.1
      appearance RustyMetal {
      }
    }
    # Cross braces
    Box {
      translation 13.5 5 2
      rotation 0 0 1 0.785
      name "scaffold_brace_1"
      size 4.24 0.05 0.05
      appearance BrushedAluminium {
      }
    }
    Box {
      translation 13.5 8 2
      rotation 0 0 1 -0.785
      name "scaffold_brace_2"
      size 4.24 0.05 0.05
      appearance BrushedAluminium {
      }
    }
  ]
}

# Construction Crane (Simplified)
DEF CONSTRUCTION_CRANE Group {
  children [
    # Crane Base
    Cylinder {
      translation -8 8 2
      name "crane_base"
      height 4
      radius 1
      appearance Concrete {
        colorOverride 0.8 0.8 0.7
      }
    }
    # Crane Mast
    Cylinder {
      translation -8 8 8
      name "crane_mast"
      height 12
      radius 0.3
      appearance BrushedAluminium {
      }
    }
    # Crane Jib
    Box {
      translation -3 8 14
      name "crane_jib"
      size 10 0.3 0.5
      appearance BrushedAluminium {
      }
    }
    # Counter Jib
    Box {
      translation -13 8 14
      name "crane_counter_jib"
      size 5 0.3 0.5
      appearance BrushedAluminium {
      }
    }
  ]
}

# Excavator (Simplified)
DEF EXCAVATOR Group {
  children [
    # Main body
    Box {
      translation -12 -8 1
      rotation 0 0 1 0.5
      name "excavator_body"
      size 4 2 2
      appearance RustyMetal {
        colorOverride 1 0.8 0
      }
    }
    # Arm
    Box {
      translation -10 -6 1.5
      rotation 0 0 1 0.3
      name "excavator_arm"
      size 3 0.3 0.3
      appearance RustyMetal {
        colorOverride 1 0.8 0
      }
    }
    # Bucket
    Box {
      translation -8.5 -5 1.2
      rotation 0 0 1 0.8
      name "excavator_bucket"
      size 1 0.8 0.5
      appearance RustyMetal {
        colorOverride 0.6 0.6 0.6
      }
    }
    # Tracks
    Box {
      translation -12 -8 0.3
      name "excavator_track_left"
      size 4 0.5 0.6
      appearance RustyMetal {
        colorOverride 0.2 0.2 0.2
      }
    }
    Box {
      translation -12 -7 0.3
      name "excavator_track_right"
      size 4 0.5 0.6
      appearance RustyMetal {
        colorOverride 0.2 0.2 0.2
      }
    }
  ]
}

# Tool Shed
DEF TOOL_SHED Group {
  children [
    Box {
      translation 18 15 1.5
      name "shed_walls"
      size 4 3 3
      appearance CorrugatedMetal {
        colorOverride 0.7 0.5 0.3
      }
    }
    Box {
      translation 18 15 3.2
      name "shed_roof"
      size 4.5 3.5 0.2
      appearance CorrugatedMetal {
        colorOverride 0.5 0.3 0.2
      }
    }
  ]
}

# Lighting for realistic environment
SpotLight {
  attenuation 0 0 1
  beamWidth 0.8
  cutOffAngle 1.2
  direction 0 -1 -1
  intensity 2
  location 10 10 8
  radius 20
  castShadows TRUE
}

SpotLight {
  attenuation 0 0 1
  beamWidth 0.8
  cutOffAngle 1.2
  direction 0 -1 -1
  intensity 2
  location -10 -10 8
  radius 20
  castShadows TRUE
}

SpotLight {
  attenuation 0 0 1
  beamWidth 0.8
  cutOffAngle 1.2
  direction 0 -1 -1
  intensity 1.5
  location 0 0 12
  radius 25
  castShadows TRUE
}

# Physics supervisor for boundary enforcement
Robot {
  name "boundary_supervisor"
  controller "boundary_supervisor"
  supervisor TRUE
  children [
    Emitter {
      name "boundary_emitter"
    }
  ]
} 